code/game/objects/explosion.dm

```diff
@@ -1,94 +1,270 @@
-		if(Debug2) to_world_log("## DEBUG: Explosion([x0],[y0],[z0])(d[devastation_range],h[heavy_impact_range],l[light_impact_range]): Took [took] seconds.")
+	if(adminlog)
+		log_and_message_admins("Explosion with size ([devastation_range], [heavy_impact_range], [light_impact_range]) in area [epicenter.loc.name] ([epicenter.x],[epicenter.y],[epicenter.z]) (<A HREF='?_src_=holder;adminplayerobservecoodjump=1;X=[epicenter.x];Y=[epicenter.y];Z=[epicenter.z]'>JMP</a>)")
 
-		sleep(8)
+	var/approximate_intensity = (devastation_range * 3) + (heavy_impact_range * 2) + light_impact_range
+	// Large enough explosion. For performance reasons, powernets will be rebuilt manually
+	if(!defer_powernet_rebuild && (approximate_intensity > 25))
+		defer_powernet_rebuild = 1
 
+	if(heavy_impact_range > 1)
+		var/datum/effect/system/explosion/E = new/datum/effect/system/explosion()
+		E.set_up(epicenter)
+		E.start()
+
+	var/x0 = epicenter.x
+	var/y0 = epicenter.y
+	var/z0 = epicenter.z
+	for(var/turf/T in trange(max_range, epicenter))
+		var/dist = sqrt((T.x - x0)**2 + (T.y - y0)**2)
+		if(dist < devastation_range)
+			dist = 1
+		else if(dist < heavy_impact_range)
+			dist = 2
+		else if(dist < light_impact_range)
+			dist = 3
+		else
+			continue
+
+		var/throw_dist = 9/dist
+		T.ex_act(dist)
+		if(!T)
+			T = locate(x0,y0,z0)
+		if(T)
+			var/throw_target = get_edge_target_turf(T, get_dir(epicenter,T))
+			for(var/atom_movable in T.contents)
+				var/atom/movable/AM = atom_movable
+				if(AM && AM.simulated && !T.protects_atom(AM))
+					AM.ex_act(dist)
+					if(!QDELETED(AM) && !AM.anchored)
+						addtimer(CALLBACK(AM, /atom/movable/.proc/throw_at, throw_target, throw_dist, throw_dist), 0)
+
+	var/took = (world.timeofday-start_time)/10
+	if(Debug2) 
+		to_world_log("## DEBUG: Explosion([x0],[y0],[z0])(d[devastation_range],h[heavy_impact_range],l[light_impact_range]): Took [took] seconds.")
 	return 1
 
+#define EXPLFX_BOTH 3
+#define EXPLFX_SOUND 2
+#define EXPLFX_SHAKE 1
+#define EXPLFX_NONE 0
+
+// All the vars used on the turf should be on unsimulated turfs too, we just don't care about those generally.
+#define SEARCH_DIR(dir) \
+	search_direction = dir;\
+	search_turf = get_step(current_turf, search_direction);\
+	if (istype(search_turf, /turf/simulated)) {\
+		turf_queue += search_turf;\
+		dir_queue += search_direction;\
+		power_queue += current_power;\
+	}
+
+/proc/explosion_iter(turf/epicenter, power, z_transfer)
+	set waitfor=0
+	if(power <= 0)
+		return
+	epicenter = get_turf(epicenter)
+	if(!epicenter)
+		return
+	message_admins("Explosion with size ([power]) in area [epicenter.loc.name] ([epicenter.x],[epicenter.y],[epicenter.z])")
+	log_game("Explosion with size ([power]) in area [epicenter.loc.name] ")
+	log_debug("iexpl: Beginning discovery phase.")
+	var/time = world.time
```

Comment:
Should use `REALTIMEOFDAY` for these so it times properly when it runs within a tick.